src/cargo/ops/tree/graph.rs

```diff
@@ -0,0 +1,562 @@
+    }
+
+    fn add_edge(&mut self, kind: Edge, index: usize) {
+        let indexes = self.0.entry(kind).or_default();
+        if !indexes.contains(&index) {
+            indexes.push(index)
+        }
+    }
+}
+
+/// A graph of dependencies.
+pub struct Graph<'a> {
+    nodes: Vec<Node>,
+    edges: Vec<Edges>,
+    /// Index maps a node to an index, for fast lookup.
+    index: HashMap<Node, usize>,
+    /// Map for looking up packages.
+    package_map: HashMap<PackageId, &'a Package>,
+    /// Set of indexes of feature nodes that were added via the command-line.
+    ///
+    /// For example `--features foo` will mark the "foo" node here.
+    cli_features: HashSet<usize>,
+    /// Map of dependency names, used for building internal feature map for
+    /// dep_name/feat_name syntax.
+    ///
+    /// Key is the index of a package node, value is a map of dep_name to a
+    /// set of `(pkg_node_index, is_optional)`.
+    dep_name_map: HashMap<usize, HashMap<InternedString, HashSet<(usize, bool)>>>,
+}
+
+impl<'a> Graph<'a> {
+    fn new(package_map: HashMap<PackageId, &'a Package>) -> Graph<'a> {
+        Graph {
+            nodes: Vec::new(),
+            edges: Vec::new(),
+            index: HashMap::new(),
+            package_map,
+            cli_features: HashSet::new(),
+            dep_name_map: HashMap::new(),
+        }
+    }
+
+    /// Adds a new node to the graph, returning its new index.
+    fn add_node(&mut self, node: Node) -> usize {
+        let from_index = self.nodes.len();
+        self.nodes.push(node);
+        self.edges.push(Edges::new());
+        self.index
+            .insert(self.nodes[from_index].clone(), from_index);
+        from_index
+    }
+
+    /// Returns a list of nodes the given node index points to for the given kind.
+    ///
+    /// Returns None if there are none.
+    pub fn connected_nodes(&self, from: usize, kind: &Edge) -> Option<Vec<usize>> {
+        match self.edges[from].0.get(kind) {
+            Some(indexes) => {
+                // Created a sorted list for consistent output.
+                let mut indexes = indexes.clone();
+                indexes.sort_unstable_by(|a, b| self.nodes[*a].cmp(&self.nodes[*b]));
+                Some(indexes)
+            }
+            None => None,
+        }
+    }
+
+    /// Gets a node by index.
+    pub fn node(&self, index: usize) -> &Node {
+        &self.nodes[index]
+    }
+
+    /// Given a slice of PackageIds, returns the indexes of all nodes that match.
+    pub fn indexes_from_ids(&self, package_ids: &[PackageId]) -> Vec<usize> {
+        let mut result: Vec<(&Node, usize)> = self
+            .nodes
+            .iter()
+            .enumerate()
+            .filter(|(_i, node)| match node {
+                Node::Package { package_id, .. } => package_ids.contains(package_id),
+                _ => false,
+            })
+            .map(|(i, node)| (node, i))
+            .collect();
+        result.sort_unstable();
```

Comment:
Could this have a comment explaining why it's sorting? (and why unstable is ok here vs a normal `sort()`?)